### ISSUE TITLE
Gateway is unable to convert the filters passed in OData URI

### ISSUE THREAD

**Issue by amritanshukr** [SAP/cloud-sdk#1197]
## The gateway is unable to convert the filters passed in OData URI to ABAP Internal Tables.

Important information:
Generated By Cloud SDK : Filter is not Working :

(TimeSheetDate%20ge%20datetime'2022-12-31T00:00:00')%20and%20(TimeSheetDate%20le%20datetime'2023-01-04T23:59:59.999999999')

 

Filter is Working : Accepted Format by Gateway:

(TimeSheetDate%20ge%20datetime%272022-12-31T00:00:00%27%20and%20TimeSheetDate%20le%20datetime%272023-01-04T23:59:59%27)

Position of Brackets is  ()and()   vs     (  and ) 
Gateway is not parsing it.

Code : 
```java
return manageWorkforceTimesheetService.getAllTimeSheetEntry()
            .select(TimeSheetEntry.ALL_FIELDS) 
          .filter(TimeSheetEntry.PERSON_WORK_AGREEMENT_EXTERNAL_ID.eq("EMPLOYEE"))       
          .filter(TimeSheetEntry.TIME_SHEET_DATE.ge(from).and(TimeSheetEntry.TIME_SHEET_DATE.le(to))) // not working
           .executeRequest(erpConfigContext);
```
 

API : API_MANAGE_WORKFORCE_TIMESHEET

Class : DefaultWorkforceTimesheetService implements WorkforceTimesheetService

Maven: 

<dependency>
                <groupId>com.sap.cloud.sdk.s4hana</groupId>
                <artifactId>s4hana-api-odata</artifactId>
                <version>4.5.0</version>
 </dependency>

 
### Impact / Priority

Affected development phase: e.g.  Development

## Error Message

```
400 status code returned
```

## Project Details

https://myXXXXX-api.s4hana.ondemand.com//sap/opu/odata/sap/API_MANAGE_WORKFORCE_TIMESHEET/TimeSheetEntryCollection?$select=*&$filter=PersonWorkAgreementExternalID%20eq%20%27EMPLOYEE%27%20and%20(TimeSheetDate%20ge%20datetime%272022-12-31T00:00:00%27%20and%20TimeSheetDate%20le%20datetime%272023-01-04T23:59:59%27)%20and%20TimeSheetStatus%20eq%20%2730%27

- SAP Cloud SDK Version: <your SAP Cloud SDK version here>

- Link to GitHub repository: <your link here>
- Project type, for example:
  - [ ] CAP Project
  - [x] SAP Cloud SDK Maven Archetype
  - [ ] None of the above:
- Platform:
  - [x] Cloud Foundry
  - [ ] Neo
  - [ ] None of the above:

---

### Checklist

- [ ] Checked out the [documentation](https://sap.github.io/cloud-sdk/docs/java/sdk-java-troubleshooting-frequent-problems) and [Stack Overflow](https://stackoverflow.com/questions/tagged/sap-cloud-sdk)
- [ ] Description provided with all relevant information
- [ ] Attached debug logs
- [ ] Attached dependency tree
- [ x] Provided the SAP Cloud SDK version & link to relevant source code

**Comment by Johannes-Schneider** [comment]
Hi @amritanshukr,

thanks for reaching out to us!

I played around with the [Workforce Timesheet Sandbox Service](https://api.sap.com/api/API_MANAGE_WORKFORCE_TIMESHEET/overview) and was able to reproduce the issue you're observing.

As it seems, the Cloud SDK is sending too many digits for the nanoseconds of the `EDM.dateTime` object.
The [OData V2 specification](https://www.odata.org/documentation/odata-version-2-0/overview/) - Chapter 6. Primitive Data Types - states the following:
> _datetime'yyyy-mm-ddThh:mm[:ss[.fffffff]]'_

![image](https://user-images.githubusercontent.com/9248711/210790416-c697efd1-4a9a-436b-a316-8f261a35d6fe.png)

So the nanoseconds are supposed to be up to **7** digits.
In the request that is assembled by the Cloud SDK, the nanoseconds consists of **9** digits, which is ultimately causing the issue - the bracket position does not matter as they are semantically equivalent.

As a consequence, I'll treat your issue as a bug report.
To workaround the issue temporarily, while we are working on a permanent solution, could you please try remove the nanoseconds part of your datetime objects?

Best regards,
Johannes

**Comment by jjtang1985** [comment]
Hi @amritanshukr ,

Do you have a dedicated timeline for your project?

**Comment by amritanshukr** [comment]
Hi @jjtang1985,

I am releasing my project in last week of Jan. Its a sample, so I will update it once you have fix ready.  
Let me know when you have fixed version ready. 

thanks,
amritanshu

**Comment by jjtang1985** [comment]
Hi @amritanshukr ,

It would be nice to understand the demo timeline for better planning from our side.

Thanks.

**Comment by Johannes-Schneider** [comment]
Hey @amritanshukr,

I spend more time playing around with the sandbox service and also used one of our internal S/4HANA Cloud systems for comparison and sanity checking.

You can find the requests below:
<details>
<summary><code>Requests.http</code></summary>

**Note**: These requests were sent using the [VS Code Rest Client Plugin](https://marketplace.visualstudio.com/items?itemName=humao.rest-client).

```
// *********************************************
// S/4HANA Cloud configuration
@s4Auth = Basic <user>:<pass>
@s4Url = https://<cloud-system>/sap/opu/odata/sap/API_BUSINESS_PARTNER/A_BusinessPartner?$select=CreationDate,FirstName&$inlinecount=allpages&$count&

@s4FilterA = CreationDate ge datetime'2022-01-01T00:00:00'
@s4FilterB = CreationDate le datetime'2024-01-01T00:00:00'
@s4FilterC = FirstName eq 'SAP'

// *********************************************
// Sandbox configuration

@sandboxKey = <api-key>
@sandboxUrl = https://sandbox.api.sap.com/s4hanacloud/sap/opu/odata/sap/API_MANAGE_WORKFORCE_TIMESHEET/TimeSheetEntryCollection?$select=TimeSheetDate,PersonWorkAgreementExternalID&$inlinecount=allpages&$count&

@sandboxFilterA = TimeSheetDate ge datetime'2017-01-01T00:00:00'
@sandboxFilterB = TimeSheetDate le datetime'2017-02-01T00:00:00'
@sandboxFilterC = PersonWorkAgreementExternalID eq 'PROJFINEMP'

// *********************************************
// c

GET {{sandboxUrl}}$filter={{sandboxFilterC}}
APIKey: {{sandboxKey}}
Accept: application/json
// https://sandbox.api.sap.com/s4hanacloud/sap/opu/odata/sap/API_MANAGE_WORKFORCE_TIMESHEET/TimeSheetEntryCollection?$select=TimeSheetDate,PersonWorkAgreementExternalID&$inlinecount=allpages&$count&$filter=PersonWorkAgreementExternalID%20eq%20'PROJFINEMP'
// --> 200, 35 results
###

GET {{s4Url}}$filter={{s4FilterC}}
Authorization: {{s4Auth}}
Accept: application/json
// https://<cloud-system>/sap/opu/odata/sap/API_BUSINESS_PARTNER/A_BusinessPartner?$select=CreationDate,FirstName&$inlinecount=allpages&$count&$filter=FirstName%20eq%20'SAP'
// --> 200, 4 results
###

// *********************************************
// a and b and c

GET {{sandboxUrl}}$filter={{sandboxFilterA}} and {{sandboxFilterB}} and {{sandboxFilterC}}
APIKey: {{sandboxKey}}
Accept: application/json
// https://sandbox.api.sap.com/s4hanacloud/sap/opu/odata/sap/API_MANAGE_WORKFORCE_TIMESHEET/TimeSheetEntryCollection?$select=TimeSheetDate,PersonWorkAgreementExternalID&$inlinecount=allpages&$count&$filter=TimeSheetDate%20ge%20datetime'2017-01-01T00:00:00'%20and%20TimeSheetDate%20le%20datetime'2017-02-01T00:00:00'%20and%20PersonWorkAgreementExternalID%20eq%20'PROJFINEMP'
// --> 200, 7 results
###

GET {{s4Url}}$filter={{s4FilterA}} and {{s4FilterB}} and {{s4FilterC}}
Authorization: {{s4Auth}}
Accept: application/json
// https://<cloud-system>/sap/opu/odata/sap/API_BUSINESS_PARTNER/A_BusinessPartner?$select=CreationDate,FirstName&$inlinecount=allpages&$count&$filter=CreationDate%20ge%20datetime'2022-01-01T00:00:00'%20and%20CreationDate%20le%20datetime'2024-01-01T00:00:00'%20and%20FirstName%20eq%20'SAP'
// --> 200, 3 results
###

// *********************************************
// (a) and (b) and (c)

GET {{sandboxUrl}}$filter=({{sandboxFilterA}}) and ({{sandboxFilterB}}) and ({{sandboxFilterC}})
APIKey: {{sandboxKey}}
Accept: application/json
// https://sandbox.api.sap.com/s4hanacloud/sap/opu/odata/sap/API_MANAGE_WORKFORCE_TIMESHEET/TimeSheetEntryCollection?$select=TimeSheetDate,PersonWorkAgreementExternalID&$inlinecount=allpages&$count&$filter=(TimeSheetDate%20ge%20datetime'2017-01-01T00:00:00')%20and%20(TimeSheetDate%20le%20datetime'2017-02-01T00:00:00')%20and%20(PersonWorkAgreementExternalID%20eq%20'PROJFINEMP')
// --> 200, 7 results
###

GET {{s4Url}}$filter=({{s4FilterA}}) and ({{s4FilterB}}) and ({{s4FilterC}})
Authorization: {{s4Auth}}
Accept: application/json
// https://<cloud-system>/sap/opu/odata/sap/API_BUSINESS_PARTNER/A_BusinessPartner?$select=CreationDate,FirstName&$inlinecount=allpages&$count&$filter=(CreationDate%20ge%20datetime'2022-01-01T00:00:00')%20and%20(CreationDate%20le%20datetime'2024-01-01T00:00:00')%20and%20(FirstName%20eq%20'SAP')
// --> 200, 3 results
###

// *********************************************
// ((a) and (b) and (c))

GET {{sandboxUrl}}$filter=(({{sandboxFilterA}}) and ({{sandboxFilterB}}) and ({{sandboxFilterC}}))
APIKey: {{sandboxKey}}
Accept: application/json
// https://sandbox.api.sap.com/s4hanacloud/sap/opu/odata/sap/API_MANAGE_WORKFORCE_TIMESHEET/TimeSheetEntryCollection?$select=TimeSheetDate,PersonWorkAgreementExternalID&$inlinecount=allpages&$count&$filter=((TimeSheetDate%20ge%20datetime'2017-01-01T00:00:00')%20and%20(TimeSheetDate%20le%20datetime'2017-02-01T00:00:00')%20and%20(PersonWorkAgreementExternalID%20eq%20'PROJFINEMP'))
// --> 200, 7 results
###

GET {{s4Url}}$filter=(({{s4FilterA}}) and ({{s4FilterB}}) and ({{s4FilterC}}))
Authorization: {{s4Auth}}
Accept: application/json
// https://<cloud-system>/sap/opu/odata/sap/API_BUSINESS_PARTNER/A_BusinessPartner?$select=CreationDate,FirstName&$inlinecount=allpages&$count&$filter=((CreationDate%20ge%20datetime'2022-01-01T00:00:00')%20and%20(CreationDate%20le%20datetime'2024-01-01T00:00:00')%20and%20(FirstName%20eq%20'SAP'))
// --> 200, 3 results
###

// *********************************************
// (a and b) and c

GET {{sandboxUrl}}$filter=({{sandboxFilterA}} and {{sandboxFilterB}}) and {{sandboxFilterC}}
APIKey: {{sandboxKey}}
Accept: application/json
// https://sandbox.api.sap.com/s4hanacloud/sap/opu/odata/sap/API_MANAGE_WORKFORCE_TIMESHEET/TimeSheetEntryCollection?$select=TimeSheetDate,PersonWorkAgreementExternalID&$inlinecount=allpages&$count&$filter=(TimeSheetDate%20ge%20datetime'2017-01-01T00:00:00'%20and%20TimeSheetDate%20le%20datetime'2017-02-01T00:00:00')%20and%20PersonWorkAgreementExternalID%20eq%20'PROJFINEMP'
// --> 200, 7 results
###

GET {{s4Url}}$filter=({{s4FilterA}} and {{s4FilterB}}) and {{s4FilterC}}
Authorization: {{s4Auth}}
Accept: application/json
// https://<cloud-system>/sap/opu/odata/sap/API_BUSINESS_PARTNER/A_BusinessPartner?$select=CreationDate,FirstName&$inlinecount=allpages&$count&$filter=(CreationDate%20ge%20datetime'2022-01-01T00:00:00'%20and%20CreationDate%20le%20datetime'2024-01-01T00:00:00')%20and%20FirstName%20eq%20'SAP'
// --> 200, 3 results
###

// *********************************************
// (a) and (b) and c

GET {{sandboxUrl}}$filter=({{sandboxFilterA}}) and ({{sandboxFilterB}}) and {{sandboxFilterC}}
APIKey: {{sandboxKey}}
Accept: application/json
// https://sandbox.api.sap.com/s4hanacloud/sap/opu/odata/sap/API_MANAGE_WORKFORCE_TIMESHEET/TimeSheetEntryCollection?$select=TimeSheetDate,PersonWorkAgreementExternalID&$inlinecount=allpages&$count&$filter=(TimeSheetDate%20ge%20datetime'2017-01-01T00:00:00')%20and%20(TimeSheetDate%20le%20datetime'2017-02-01T00:00:00')%20and%20PersonWorkAgreementExternalID%20eq%20'PROJFINEMP'
// --> 200, 7 results
###

GET {{s4Url}}$filter=({{s4FilterA}}) and ({{s4FilterB}}) and {{s4FilterC}}
Authorization: {{s4Auth}}
Accept: application/json
// https://<cloud-system>/sap/opu/odata/sap/API_BUSINESS_PARTNER/A_BusinessPartner?$select=CreationDate,FirstName&$inlinecount=allpages&$count&$filter=(CreationDate%20ge%20datetime'2022-01-01T00:00:00')%20and%20(CreationDate%20le%20datetime'2024-01-01T00:00:00')%20and%20FirstName%20eq%20'SAP'
// --> 200, 3 results
###

// *********************************************
// ((a) and (b)) and c

GET {{sandboxUrl}}$filter=(({{sandboxFilterA}}) and ({{sandboxFilterB}})) and {{sandboxFilterC}}
APIKey: {{sandboxKey}}
Accept: application/json
// https://sandbox.api.sap.com/s4hanacloud/sap/opu/odata/sap/API_MANAGE_WORKFORCE_TIMESHEET/TimeSheetEntryCollection?$select=TimeSheetDate,PersonWorkAgreementExternalID&$inlinecount=allpages&$count&$filter=((TimeSheetDate%20ge%20datetime'2017-01-01T00:00:00')%20and%20(TimeSheetDate%20le%20datetime'2017-02-01T00:00:00'))%20and%20PersonWorkAgreementExternalID%20eq%20'PROJFINEMP'
// --> 200, 7 results
###

GET {{s4Url}}$filter=(({{s4FilterA}}) and ({{s4FilterB}})) and {{s4FilterC}}
Authorization: {{s4Auth}}
Accept: application/json
// https://<cloud-system>/sap/opu/odata/sap/API_BUSINESS_PARTNER/A_BusinessPartner?$select=CreationDate,FirstName&$inlinecount=allpages&$count&$filter=((CreationDate%20ge%20datetime'2022-01-01T00:00:00')%20and%20(CreationDate%20le%20datetime'2024-01-01T00:00:00'))%20and%20FirstName%20eq%20'SAP'
// --> 200, 3 results
###

// *********************************************
// c and (a) and (b)

GET {{sandboxUrl}}$filter={{sandboxFilterC}} and ({{sandboxFilterA}}) and ({{sandboxFilterB}})
APIKey: {{sandboxKey}}
Accept: application/json
// https://sandbox.api.sap.com/s4hanacloud/sap/opu/odata/sap/API_MANAGE_WORKFORCE_TIMESHEET/TimeSheetEntryCollection?$select=TimeSheetDate,PersonWorkAgreementExternalID&$inlinecount=allpages&$count&$filter=PersonWorkAgreementExternalID%20eq%20'PROJFINEMP'%20and%20(TimeSheetDate%20ge%20datetime'2017-01-01T00:00:00')%20and%20(TimeSheetDate%20le%20datetime'2017-02-01T00:00:00')
// --> 200, 992 results --> wrong result!
###

GET {{s4Url}}$filter={{s4FilterC}} and ({{s4FilterA}}) and ({{s4FilterB}})
Authorization: {{s4Auth}}
Accept: application/json
// https://<cloud-system>/sap/opu/odata/sap/API_BUSINESS_PARTNER/A_BusinessPartner?$select=CreationDate,FirstName&$inlinecount=allpages&$count&$filter=FirstName%20eq%20'SAP'%20and%20(CreationDate%20ge%20datetime'2022-01-01T00:00:00')%20and%20(CreationDate%20le%20datetime'2024-01-01T00:00:00')
// --> 200, 3 results
###

// *********************************************
// c and (a and b)

GET {{sandboxUrl}}$filter={{sandboxFilterC}} and ({{sandboxFilterA}} and {{sandboxFilterB}})
APIKey: {{sandboxKey}}
Accept: application/json
// https://sandbox.api.sap.com/s4hanacloud/sap/opu/odata/sap/API_MANAGE_WORKFORCE_TIMESHEET/TimeSheetEntryCollection?$select=TimeSheetDate,PersonWorkAgreementExternalID&$inlinecount=allpages&$count&$filter=PersonWorkAgreementExternalID%20eq%20'PROJFINEMP'%20and%20(TimeSheetDate%20ge%20datetime'2017-01-01T00:00:00'%20and%20TimeSheetDate%20le%20datetime'2017-02-01T00:00:00')
// --> 200, 7 results 
###

GET {{s4Url}}$filter={{s4FilterC}} and ({{s4FilterA}} and {{s4FilterB}})
Authorization: {{s4Auth}}
Accept: application/json
// https://<cloud-system>/sap/opu/odata/sap/API_BUSINESS_PARTNER/A_BusinessPartner?$select=CreationDate,FirstName&$inlinecount=allpages&$count&$filter=FirstName%20eq%20'SAP'%20and%20(CreationDate%20ge%20datetime'2022-01-01T00:00:00'%20and%20CreationDate%20le%20datetime'2024-01-01T00:00:00')
// --> 200, 3 results
###
```
</details>

Following observations are most noteworthy:
<details>

```
// *********************************************
// (a) and (b) and c

GET {{sandboxUrl}}$filter=({{sandboxFilterA}}) and ({{sandboxFilterB}}) and {{sandboxFilterC}}
APIKey: {{sandboxKey}}
Accept: application/json
// https://sandbox.api.sap.com/s4hanacloud/sap/opu/odata/sap/API_MANAGE_WORKFORCE_TIMESHEET/TimeSheetEntryCollection?$select=TimeSheetDate,PersonWorkAgreementExternalID&$inlinecount=allpages&$count&$filter=(TimeSheetDate%20ge%20datetime'2017-01-01T00:00:00')%20and%20(TimeSheetDate%20le%20datetime'2017-02-01T00:00:00')%20and%20PersonWorkAgreementExternalID%20eq%20'PROJFINEMP'
// --> 200, 7 results
###

// *********************************************
// c and (a) and (b)

GET {{sandboxUrl}}$filter={{sandboxFilterC}} and ({{sandboxFilterA}}) and ({{sandboxFilterB}})
APIKey: {{sandboxKey}}
Accept: application/json
// https://sandbox.api.sap.com/s4hanacloud/sap/opu/odata/sap/API_MANAGE_WORKFORCE_TIMESHEET/TimeSheetEntryCollection?$select=TimeSheetDate,PersonWorkAgreementExternalID&$inlinecount=allpages&$count&$filter=PersonWorkAgreementExternalID%20eq%20'PROJFINEMP'%20and%20(TimeSheetDate%20ge%20datetime'2017-01-01T00:00:00')%20and%20(TimeSheetDate%20le%20datetime'2017-02-01T00:00:00')
// --> 200, 992 results --> wrong result!
###

// *********************************************
// c and (a and b)

GET {{sandboxUrl}}$filter={{sandboxFilterC}} and ({{sandboxFilterA}} and {{sandboxFilterB}})
APIKey: {{sandboxKey}}
Accept: application/json
// https://sandbox.api.sap.com/s4hanacloud/sap/opu/odata/sap/API_MANAGE_WORKFORCE_TIMESHEET/TimeSheetEntryCollection?$select=TimeSheetDate,PersonWorkAgreementExternalID&$inlinecount=allpages&$count&$filter=PersonWorkAgreementExternalID%20eq%20'PROJFINEMP'%20and%20(TimeSheetDate%20ge%20datetime'2017-01-01T00:00:00'%20and%20TimeSheetDate%20le%20datetime'2017-02-01T00:00:00')
// --> 200, 7 results 
###

// *********************************************
// (a and b) and c

GET {{sandboxUrl}}$filter=({{sandboxFilterA}} and {{sandboxFilterB}}) and {{sandboxFilterC}}
APIKey: {{sandboxKey}}
Accept: application/json
// https://sandbox.api.sap.com/s4hanacloud/sap/opu/odata/sap/API_MANAGE_WORKFORCE_TIMESHEET/TimeSheetEntryCollection?$select=TimeSheetDate,PersonWorkAgreementExternalID&$inlinecount=allpages&$count&$filter=(TimeSheetDate%20ge%20datetime'2017-01-01T00:00:00'%20and%20TimeSheetDate%20le%20datetime'2017-02-01T00:00:00')%20and%20PersonWorkAgreementExternalID%20eq%20'PROJFINEMP'
// --> 200, 7 results
###
```

</details>

To summarize, following filters yield following results:

* 🟢 `(a and b) and c` 
* 🟢 `(a) and (b) and c`
* 🟢 `c and (a and b)`
* 🔴 `c and (a) and (b)`

Luckily, this is only an issue in the sandbox environment. Our S/4 Cloud system always returned the correct result, no matter how the `$filter` was assembled.

To get back to your issue:
Can you try switching up the order of the `.filter` operations like so:

```java
return manageWorkforceTimesheetService
        .getAllTimeSheetEntry()
        .select(TimeSheetEntry.ALL_FIELDS)
        .filter(TimeSheetEntry.TIME_SHEET_DATE.ge(from).and(TimeSheetEntry.TIME_SHEET_DATE.le(to)))
        .filter(TimeSheetEntry.PERSON_WORK_AGREEMENT_EXTERNAL_ID.eq("EMPLOYEE"))
        .executeRequest(erpConfigContext);
```

Best regards,
Johannes

**Comment by newtork** [comment]
Hi @amritanshukr,

As we found out the underlying issue is with the sandbox server (and not with a productive system). We have not derived any additional steps apart from the [suggested workaround above](https://github.com/SAP/cloud-sdk/issues/1197#issuecomment-1385589555). Did this solve the issue for you? Or do you require any additional assistance?

Otherwise, I would go ahead and close the issue.
Kind regards,
Alexander

**Comment by amritanshukr** [comment]
Hi @newtork,

The work around worked in the test system. We can close the issue, I did raise ticket for sandbox system but it was turned down by dispatcher.
thanks a ton!! 👍 

kind regards,
amritanshu
